### ISSUE TITLE
How to get/set PWM/channel values in C/C++?

### ISSUE THREAD

**Issue by renne** [arendst/Sonoff-Tasmota#3422]
Hi,

I'm working on a closed-loop light-control function which needs direct C/C++ access without MQTT parsing overhead to the PWM1-5/channel1-5 functions and variables. Where can I find them in the code?

**Comment by andrethomas** [comment]
xdrv_04_light.ino, I think.

**Comment by andrethomas** [comment]
Its the only place I can find that has AnalogWrite's in it...

**Comment by renne** [comment]
In that case I need to know to which PINs PWM1-5 are mapped. Do you have any idea?

**Comment by andrethomas** [comment]
I think the mapping will depend on how the device is configured in the web menu, or do you perhaps mean from wirish...

**Comment by renne** [comment]
That's the point. I'm currently working on extending the APDS-9960 code with closed-loop light control. As there are different devices with different GPIO mappings I need a generic way to set the PWM values or at least query the GPIO mapping.

**Comment by andrethomas** [comment]
Then I think the best would be to interact with xdrv_04_light.ino to add compiler directives on whether or not to use data from APDS-9960 through a shared global variable perhaps.

**Comment by ascillato2** [comment]
Hi,

Have you managed to solve your issue?

**Comment by ascillato2** [comment]
Yes, exactly

**Comment by ascillato2** [comment]
Hi,

Have you managed to solve your issue?

**Comment by renne** [comment]
I'm currently waiting on the Broadcom/EBV support for more information about the color-space of the APDS-9960. Until then I have to put the closed-loop light control on hold.

**Comment by ascillato** [comment]
Ok, no problem, but the question of this issue has been answered? There is something else we can help?

**Comment by renne** [comment]
In theory, yes. ;-)
